File: src/main/java/jp/aoichaan0513/A_TosoGame_Live/Listeners/onQuit.java
package jp.aoichaan0513.A_TosoGame_Live.Listeners;

import jp.aoichaan0513.A_TosoGame_Live.API.MainAPI;
import jp.aoichaan0513.A_TosoGame_Live.API.Manager.MissionManager;
import jp.aoichaan0513.A_TosoGame_Live.API.Scoreboard.Scoreboard;
import jp.aoichaan0513.A_TosoGame_Live.API.Scoreboard.Teams;
import jp.aoichaan0513.A_TosoGame_Live.API.TosoGameAPI;
import jp.aoichaan0513.A_TosoGame_Live.Main;
import org.bukkit.Bukkit;
import org.bukkit.ChatColor;
import org.bukkit.Material;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.player.PlayerQuitEvent;
import org.bukkit.inventory.meta.ItemMeta;
import org.bukkit.scoreboard.DisplaySlot;

public class onQuit implements Listener {

    @EventHandler
    public void onQuit(PlayerQuitEvent e) {
        Player p = e.getPlayer();

        e.setQuitMessage(!MainAPI.isHidePlayer(p) ? ChatColor.YELLOW + "<- " + ChatColor.GOLD + p.getName() : "");

        Main.hunterShuffleSet.remove(p.getUniqueId());
        Main.tuhoShuffleSet.remove(p.getUniqueId());

        Main.opGamePlayerSet.remove(p.getUniqueId());
        onMove.zoneList.remove(p);

        if (Main.invisibleSet.contains(p.getUniqueId())) {
            for (Player player : Bukkit.getOnlinePlayers())
                player.showPlayer(p);
            Main.invisibleSet.remove(p.getUniqueId());
        }

        if (MissionManager.isBossBar())
            MissionManager.getBossBar().removePlayer(p);

        for (Player player : Bukkit.getOnlinePlayers())
            TosoGameAPI.showPlayers(player);

        if (Teams.hasJoinedTeam(Teams.OnlineTeam.TOSO_HUNTER, p)) {
            for (Player player : Bukkit.getOnlinePlayers())
                if (Teams.hasJoinedTeam(Teams.OnlineTeam.TOSO_ADMIN, player))
                    player.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "ハンターが退出しました。");
        } else if (Teams.hasJoinedTeam(Teams.OnlineTeam.TOSO_SUCCESS, p)) {
            Teams.leaveTeam(Teams.OnlineTeam.TOSO_SUCCESS, p);
            Teams.joinTeam(Teams.OnlineTeam.TOSO_PLAYER, p);

        }

        for (Player player : Bukkit.getOnlinePlayers()) {
            org.bukkit.scoreboard.Scoreboard board = Scoreboard.getBoard(player);

            if (!TosoGameAPI.isAdmin(player)) {
                if (player.getInventory().getItemInMainHand().getType() == Material.BOOK) {
                    ItemMeta itemMeta = player.getInventory().getItemInMainHand().getItemMeta();

                    if (ChatColor.stripColor(itemMeta.getDisplayName()).equals(Main.PHONE_ITEM_NAME)) {
                        board.getObjective(TosoGameAPI.Objective.SIDEBAR.getName()).setDisplaySlot(DisplaySlot.SIDEBAR);
                    } else {
                        board.clearSlot(DisplaySlot.SIDEBAR);
                    }
                } else if (player.getInventory().getItemInOffHand().getType() == Material.BOOK) {
                    ItemMeta itemMeta = player.getInventory().getItemInOffHand().getItemMeta();

                    if (ChatColor.stripColor(itemMeta.getDisplayName()).equals(Main.PHONE_ITEM_NAME)) {
                        board.getObjective(TosoGameAPI.Objective.SIDEBAR.getName()).setDisplaySlot(DisplaySlot.SIDEBAR);
                    } else {
                        board.clearSlot(DisplaySlot.SIDEBAR);
                    }
                } else {
                    board.clearSlot(DisplaySlot.SIDEBAR);
                }
            } else {
                board.getObjective(TosoGameAPI.Objective.SIDEBAR.getName()).setDisplaySlot(DisplaySlot.SIDEBAR);
            }
            return;
        }
    }
}

File: src/main/kotlin/jp/aoichaan0513/A_TosoGame_Live/Listeners/Minecraft/onQuit.kt
package jp.aoichaan0513.A_TosoGame_Live.Listeners.Minecraft

import jp.aoichaan0513.A_TosoGame_Live.API.MainAPI
import jp.aoichaan0513.A_TosoGame_Live.API.MainAPI.PrefixType
import jp.aoichaan0513.A_TosoGame_Live.API.Manager.Player.PlayerManager
import jp.aoichaan0513.A_TosoGame_Live.API.Manager.Player.VisibilityManager
import jp.aoichaan0513.A_TosoGame_Live.API.Scoreboard.Teams
import jp.aoichaan0513.A_TosoGame_Live.API.Scoreboard.Teams.OnlineTeam
import jp.aoichaan0513.A_TosoGame_Live.Main
import jp.aoichaan0513.A_TosoGame_Live.Mission.HunterZone
import jp.aoichaan0513.A_TosoGame_Live.Mission.MissionManager
import jp.aoichaan0513.A_TosoGame_Live.Mission.TimedDevice
import jp.aoichaan0513.A_TosoGame_Live.OPGame.Dice
import jp.aoichaan0513.A_TosoGame_Live.Utils.PlayerUtil
import jp.aoichaan0513.A_TosoGame_Live.Utils.isAdminTeam
import jp.aoichaan0513.A_TosoGame_Live.Utils.isHunterTeam
import jp.aoichaan0513.A_TosoGame_Live.Utils.isSuccessTeam
import org.bukkit.Bukkit
import org.bukkit.ChatColor
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.player.PlayerQuitEvent

class onQuit : Listener {

    @EventHandler
    fun onQuit(e: PlayerQuitEvent) {
        val p = e.player

        e.quitMessage = if (!PlayerManager.loadConfig(p).visibility) "${ChatColor.YELLOW}<- ${ChatColor.GOLD}${p.name}" else ""

        Main.hunterShuffleSet.remove(p.uniqueId)
        Main.tuhoShuffleSet.remove(p.uniqueId)

        Bukkit.getScheduler().runTaskLater(Main.pluginInstance, Runnable { Dice.setPlayer() }, 2)
        HunterZone.removeJoinedSet(p, true)
        TimedDevice.addFailedNumberSet(p)

        VisibilityManager.remove(p, VisibilityManager.VisibilityType.ITEM)

        MissionManager.bossBar?.removePlayer(p)

        if (p.isHunterTeam) {
            for (player in Bukkit.getOnlinePlayers().filter { it.isAdminTeam })
                player.sendMessage("${MainAPI.getPrefix(PrefixType.ERROR)}ハンターが退出しました。")
        } else if (p.isSuccessTeam) {
            Teams.joinTeam(OnlineTeam.TOSO_PLAYER, p)
        }

        PlayerUtil.setSidebar()
    }
}